### DISCUSSION TITLE
[SDK] Issue using Paymaster with `zkSync-web3.Web3Provider` and WalletConnect-v2

### DISCUSSION THREAD

**Discussion by equalogix** [zkSync-Community-Hub/zksync-developers#63] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Python and Java SDK as they are being updated.

### Issue Description

It is not possible to send transactions using a Paymaster when using WalletConnect-v2 Ethereum-Provider (tested with Metamask as the wallet), the transactions are sent to the wallet, but no approval pop-up is ever triggered and in the app an error is thrown with the generic message `{code: -32603, message: 'internal error'}`.

When executing the same code using Metamask instead of WalletConnect, both on desktop and from the MM-Browser transaction are sent successfully without throwing any errors.

When one replaces the `zkSync-web3.Web3Provider` with the `ethers.js` `Web3Provider` and one uses WalletConnect, then it throws an error for invalid key `customData` which is expected (but not desired). In this case, it also throws the error when using Metamask to connect.

We have tried to send the transactions using `contract[method]` and `sendTransaction` with the contract method/abi and paymaster encoded, in both cases the transactions fail. We have also tried `contract.populateTransaction` and `sendUncheckedTransaction` but that simply ignores the paymaster and requires the user to pay the gas fees. I have opened a separate thread for that issue, as sending transaction using the method is required when using WC on iOS. See: https://github.com/zkSync-Community-Hub/zkync-developers/discussions/53

### Expected Behavior

When a transaction is sent using a Paymaster it should open an approval pop-up in the wallet to sign a type 113 transaction and not throw any errors.

### Code Example

```
  const overrides = { 
    customData: { 
      gasPerPubdata: ethers.BigNumber.from(utils.DEFAULT_GAS_PER_PUBDATA_LIMIT),
      paymasterAddress, 
          {
            type: "General",
            innerInput: new Uint8Array(),
          }
    } 
  }

  useEffect(() => {
    if(!contractRef.current && address && signer){ 
      contractRef.current = new Contract(
        address,
        abi,
        signer
      )
    }
  },[address, abi, signer])

  useEffect(() => {
    if(gasPrice?.maxFeePerGas && gasLimit && nonce && !finalOverrides){
      setFinalOverrides(state => ({
        customData: {...overrides.customData},
        maxFeePerGas: gasPrice.maxFeePerGas,
        maxPriorityFeePerGas: gasPrice.maxPriorityFeePerGas,
        gasLimit,
        type: 113,
        nonce
      }))
    }
  },[gasPrice?.maxFeePerGas, gasLimit, overrides])

const transact = () => {
  contractRef.current[method](...params, finalOverrides)
    .then( txHash => {
      txHash.wait()
        .then( awaited =>  { /* ... do something with the result... */})
    })
}
```

**Comment by dutterbutter** [comment]
Hey there! 😊 Thanks for the super detailed info you've provided here and in #53! Sorry for the delayed response, we'll aim to be speedier next time. 🐢➡️🐇

We understand how crucial it is for WalletConnect and paymasters to work seamlessly together, so rest assured we're on the case. We will be investigating this further to find a resolution for yourself and the community. We want to give you the ease-of-use you're after! I know patience is a virtue and you've been super virtuous so far! 🙏 Thank you for that. 

I will update here as soon as we have some updates.

Again, thank you for being an awesome community member! 🙌

**Reply by equalogix** [reply]
Thank you very much. I'm available if you need more information or other assistance.

**Reply by davidonuh** [reply]
this was very helpful btw thanks

**Comment by equalogix** [comment]
By chance, I figured out what the problem was. It was a really small mistake in how WalletConnect's `EthereumProvider` was being initialized. When specifying the methods to use in the session, I used the key `methods` instead of `optionalMethods`, here are the docs explaining this:  https://docs.walletconnect.com/2.0/web/providers/ethereum#required-and-optional-methods

But the real problem remains. The issue is the error returned is always `{code: 32603, message: internal-error}` regardless of the actual error. For example, if a transaction is cancelled, I should get a code 4001 error but I still get 32603 error. The problem I just resolved could have been fixed in minutes, if I only had the correct error message.

The big question now is what is causing the error message bug? Is it `WalletConnect`, `zksync-web3` or `MetaMask` or any combination? I believe that this is a WalletConnect issue because there are other reports in their repo from devs not using zkSync of the same error message, but again the message is so generic that it is impossible to know.

Here are some examples of reports of the issue:
WC github repo:
- https://github.com/WalletConnect/walletconnect-monorepo/issues/2825

MM github repo:
- https://github.com/MetaMask/metamask-mobile/issues/6740
- https://github.com/MetaMask/metamask-mobile/issues/6715

Assuming that this is an issue with WC or MM, ideally if the zkSync team could reach out to WC/MM it could potentially get this issue prioritized.

Also this does nothing to solve my issue #53 which remains a big concern and for which I see no easy solution.

**Comment by equalogix** [comment]
Good and bad news.
Good news: Metamask/mobile released an update we are now at v7.3.0. It fixed the bug with the error messages, I am now seeing a full error message: `{code: 5000, message: '{"code":-32603,"message":"Internal JSON-RPC error.…ata":{"code":-32600,"message":"Invalid request"}}'}`

Bad news: It stopped working again. WTF!!! I feel like I'm going crazy... I didn't change anything on my end, and the everything continues 
to work fine in the MM-browser, but I can't get WalletConnect to work, it's broken both on iOS and Android.

**Comment by equalogix** [comment]
I should add, that this doesn't appear to be related to the paymaster. I have tried with and without including the custom data and neither works. The only way currently to interact with a smart contract using WC is the use `contract.populateTransaction` followed by  `sendUncheckedTransaction`, which ignores the paymaster.

**Comment by equalogix** [comment]
I figured it out, (I hope!) the issue is with WalletConnect, and the passing of the methods during the initialization of the `EthereumProvider`. The TS type for the methods object is limited to a set of specific methods, if you try to pass a method that is not explicitly declared in WC's code it throws the generic error message.

Here is the link to WC's code that shows the explicit list.
https://github.com/WalletConnect/walletconnect-monorepo/blob/dc26b396c3a1a2f266143289842d7c7695eaf3ef/providers/ethereum-provider/src/EthereumProvider.ts#L36

**Reply by dutterbutter** [reply]
Thank you for this. I intend to reproduce and document the behaviour explicitly to share with their team. Thank you for the diligent debugging here!

**Comment by equalogix** [comment]
I was planning on opening issue in the WC repo for this. This isn't really a bug, it's just that the error thrown provides you with no clue to what the problem.

**Comment by equalogix** [comment]
Guess what? The issue is not resolved. Last week when I was working on this I made a fatal assumption. When clicking the "mint token" button in my dApp MetaMask correctly displayed the transaction pop-up with type 113 transaction message to sign. I assumed that it meant that it was good, because once I've reached that point it was always good. So as not to waste testnet tokens I simply clicked cancel. But I was wrong, it wasn't good.

This morning I got back onto the project, I tried minting, and I did click submit, and the transaction failed. `internal JSON-RPC error`. I tried again a few times, it always fails, checked the paymaster balance to ensure there was sufficient funds to cover the gas fees, all is good, but transaction still fails. I then tried it the MM Browser and it worked no problem.

Since now both WC and MM provide me with the transaction message, I compared the messages and they are both identical. I really don't get what the problem could be.

Here is the message:
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/16868716/8eb1d915-dd5d-4f46-9841-a7231478bea0)

**Reply by dutterbutter** [reply, marked as answer]
It seems that WC does not support paymaster `customData` payload at this time. We are investigating a patch or a way to circumvent this for the short-term, and then hopefully a longer-term solution is an upstream fix to WC. Paymaster transactions will work fine with MM so its specific to WC. This issue is separate from the issues you faced prior and tagged accordingly.

**Reply by equalogix** [reply]
Thanks for this confirmation. It is worth mentioning that I was using the paymaster with WC without any issue when I was using WCv1, this is a bug/side-effect that was introduced with the switch to V2.

Note: I opened an issue with WC https://github.com/WalletConnect/walletconnect-monorepo/issues/3269

**Reply by lutr0** [reply]
Hi, is there any update on WC? I got to the same error trying to send a paymaster tx.
Apart from Metamask, which other wallets work?

**Reply by glitch-txs** [reply]
Hi! is this still an issue? could you please provide a minimal reproducible example that we can run?

EthereumProvider is meant to follow EIP-1163 so for custom methods we usually use Universal Provider, but we can add more to the first one if required.

**Reply by bxpana** [reply]
@lutr0 and @glitch-txs this should be working with WC now. If you run into any issues feel free to open up a new discussion

**Comment by lysumeduso** [comment]
It seems like there may be compatibility issues between the WalletConnect-v2 Ethereum-Provider and the zkSync Paymaster functionality. This could be due to differences in how the two systems handle transaction approval and custom data. To resolve this issue, you may need to investigate and potentially modify the way custom data is handled when interacting with the WalletConnect provider. Additionally, consider reaching out to the WalletConnect and zkSync support teams for further assistance and clarification on any potential compatibility issues between their systems.

**Comment by ljada37** [comment]
The issue seems to stem from the interaction between WalletConnect-v2 Ethereum-Provider and the Paymaster functionality. Transactions sent using WalletConnect do not trigger the expected approval pop-up in the wallet, leading to an error message {code: -32603, message: 'internal error'}. This behavior is not observed when using Metamask directly. Further investigation is needed to identify and resolve the root cause of this discrepancy in behavior between WalletConnect and direct Metamask usage.

**Comment by heuresfanegarnold** [comment]
It appears that there is an issue with sending transactions using a Paymaster when using WalletConnect-v2 Ethereum-Provider. The transactions are sent to the wallet, but no approval pop-up is triggered, resulting in an error with the message {code: -32603, message: 'internal error'}.

One possible solution is to ensure that the customData object is correctly formatted and includes the paymasterAddress. Additionally, it's essential to verify that the gasPerPubdata value is set appropriately.

You may also consider reaching out to the maintainers of WalletConnect-v2 Ethereum-Provider for further assistance in resolving this issue.

**Comment by yamyv** [comment]
It appears that there are issues with sending transactions using a Paymaster when using WalletConnect-v2 Ethereum-Provider, resulting in a failure to trigger the approval pop-up in the wallet. Additionally, errors with the generic message {code: -32603, message: 'internal error'} are encountered. The expected behavior is for the approval pop-up to appear without any errors. Further investigation and potential fixes may be required to resolve this issue.

**Comment by camicagochristie** [comment]
Issue: Unable to trigger approval pop-up for transactions using Paymaster when using WalletConnect-v2 Ethereum-Provider, resulting in internal error.
